src/storage/tests/setup.rs

```diff
@@ -0,0 +1,263 @@
+                status_shard: None,
+            };
+            let data_shard = collection_metadata.data_shard.clone();
+            let id = GlobalId::User(1);
+            let source_exports = BTreeMap::from([(
+                id,
+                mz_storage::types::sources::SourceExport {
+                    storage_metadata: collection_metadata.clone(),
+                    output_index: 0,
+                },
+            )]);
+
+            {
+                let _tokio_guard = tokio_runtime.enter();
+                worker.handle_storage_command(StorageCommand::CreateSources(vec![
+                    mz_storage::protocol::client::CreateSourceCommand {
+                        id,
+                        description: mz_storage::types::sources::IngestionDescription {
+                            desc: desc.clone(),
+                            ingestion_metadata: collection_metadata,
+                            source_exports,
+                            // Only used for Debezium
+                            source_imports: BTreeMap::new(),
+                            host_config: mz_storage::types::hosts::StorageHostConfig::Remote {
+                                addr: "test".to_string(),
+                            },
+                        },
+                        resume_upper: Antichain::from_elem(Timestamp::minimum()),
+                    },
+                ]));
+            }
+
+            // Run the assertions in a tokio task, so we can step the dataflow
+            // while we check the snapshot
+            let check_task = {
+                let until = until.clone();
+                (&tokio_runtime).spawn_named(|| "check_loop".to_string(), async move {
+                    loop {
+                        let (mut data_write_handle, data_read_handle) = persist_client
+                            .open::<SourceData, (), Timestamp, Diff>(data_shard.clone())
+                            .await
+                            .unwrap();
+                        if let Some(values) = until(
+                            data_write_handle.fetch_recent_upper().await.clone(),
+                            data_read_handle,
+                        )
+                        .await
+                        {
+                            return values;
+                        }
+                        tokio::time::sleep(timestamp_interval).await;
+                    }
+                })
+            };
+
+            {
+                let _tokio_guard = tokio_runtime.enter();
+                while !check_task.is_finished() {
+                    worker.timely_worker.step();
+                }
+
+                // Drop the dataflow before we move on, as we could
+                // get additional activations that cause problems.
+                worker
+                    .timely_worker
+                    .drop_dataflow(worker.timely_worker.installed_dataflows()[0]);
```

Comment:
@aljoscha I dont think this will work, as we are actually required to turn off all scheduling (even from timers) to ensure we don't panic after the tokio runtime is closed!

`AllowCompaction` doesn't actually shutdown the dataflow does it?

Reply:
it should, `AllowCompaction([])` is basically the "drop everything" message, when we receive it we take away the source/sink tokens, which makes everything shut down